### PR TITLE
support file-like objects in xarray.open_rasterio

### LINKED ISSUE #4139: [Feature request] Support file-like objects in open_rasterio
With some acrobatics, it is possible to open file-like objects to rasterio. It would be useful if xarray supported this workflow, particularly for working with cloud optimized geotiffs and fs-spec. 

#### MCVE Code Sample

```python
with open('my_data.tif', 'rb') as f:
    da = xr.open_rasterio(f)

```

#### Expected Output
DataArray -> equivalent to `xr.open_rasterio('my_data.tif')`

#### Problem Description

We only currently allow str, rasterio.DatasetReader, or rasterio.WarpedVRT as inputs to `open_rasterio`.

#### Versions

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

INSTALLED VERSIONS
------------------
commit: 2a288f6ed4286910fcf3ab9895e1e9cbd44d30b4
python: 3.8.2 | packaged by conda-forge | (default, Apr 24 2020, 07:56:27)
[Clang 9.0.1 ]
python-bits: 64
OS: Darwin
OS-release: 18.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8
libhdf5: None
libnetcdf: None

xarray: 0.15.2.dev68+gb896a68f
pandas: 1.0.4
numpy: 1.18.5
scipy: None
netCDF4: None
pydap: None
h5netcdf: None
h5py: None
Nio: None
zarr: None
cftime: None
nc_time_axis: None
PseudoNetCDF: None
rasterio: 1.1.5
cfgrib: None
iris: None
bottleneck: None
dask: 2.18.1
distributed: 2.18.0
matplotlib: None
cartopy: None
seaborn: None
numbagg: None
pint: None
setuptools: 46.1.3.post20200325
pip: 20.1
conda: None
pytest: 5.4.3
IPython: 7.13.0
sphinx: 3.0.3

</details>

xref: https://github.com/pangeo-data/pangeo-datastore/issues/109

### PR DESCRIPTION
**Pull request by jhamman**
- [x] Closes #4139
 - [x] Tests added
 - [x] Passes `isort -rc . && black . && mypy . && flake8`
 - [ ] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

cc @scottyhq and @martindurant 

xref: https://github.com/pangeo-data/pangeo-datastore/issues/109

### CODE REVIEW

File: xarray/backends/rasterio_.py [hunk 1]
```diff
@@ -20,6 +21,12 @@
 )
 
 
+def _file_object_opener(fobj, *args, **kwargs):
+    from rasterio.io import MemoryFile
+
+    return MemoryFile(fobj).open(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (jhamman) [reply xarray/backends/rasterio_.py:27]: xref: https://github.com/mapbox/rasterio/issues/977

[the remark this answers is not in the document]
Comment (martindurant) [reply xarray/backends/rasterio_.py:29]: se here, can we just add `MemoryFile` to the superclasses of `fobj`, so rasterio is tricked into thinking it's dealing with a memoryfile, but the implementation is actually the original code for the object? Or maybe subclass MemoryFile here with methods that defer to the fobj.

**Comment by jhamman** [comment]
closed via https://github.com/rasterio/rasterio/pull/2141

**Comment by rsignell-usgs** [comment]
I'd like to use this cool new rasterio/fspec functionality in xarray!  What am I doing wrong here:
https://nbviewer.org/gist/rsignell-usgs/dbf3d8e952895ca255f300790759c60f

**Comment by snowman2** [comment]
Sounds like that may need changes in rioxarray. Mind opening up an issue: https://github.com/corteva/rioxarray/issues ?

**Comment by rsignell-usgs** [comment]
Thanks @snowman2 !  Done in https://github.com/corteva/rioxarray/issues/440

**Pull request closed without merging**